components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/EmailDomainValidationHandler.java

```diff
@@ -0,0 +1,232 @@
+
+    private static final Log log = LogFactory.getLog(EmailDomainValidationHandler.class);
+    private static final String EMAIL_DOMAIN_ENABLE = "emailDomain.enable";
+    public static final String EMAIL_DOMAIN = "emailDomain";
+
+    private EmailDomainValidationHandler() {
+
+    }
+
+    private static class Holder {
+
+        private static final EmailDomainValidationHandler INSTANCE = new EmailDomainValidationHandler();
+    }
+
+    public static EmailDomainValidationHandler getInstance() {
+
+        return Holder.INSTANCE;
+    }
+
+    @Override
+    public boolean isEnabled() {
+
+        if (!super.isEnabled()) {
+            return false;
+        }
+
+        try {
+            OrganizationManager organizationManager = FrameworkServiceDataHolder.getInstance().getOrganizationManager();
+            String organizationId = organizationManager.resolveOrganizationId(CarbonContext
+                    .getThreadLocalCarbonContext().getTenantDomain());
+
+            if (organizationManager.isPrimaryOrganization(organizationId)) {
+                // Skip email domain validation since email domains cannot be mapped to primary organizations.
+                return false;
+            } else {
+                organizationId = organizationManager.getPrimaryOrganizationId(organizationId);
+            }
+
+            return isEmailDomainDiscoveryEnabled(organizationId);
+        } catch (OrganizationConfigClientException e) {
+            if (log.isDebugEnabled()) {
+                log.debug("No organization discovery configurations found for organization: " + CarbonContext
+                        .getThreadLocalCarbonContext().getTenantDomain());
+            }
+            return false;
+        } catch (OrganizationManagementException | OrganizationConfigException e) {
+            log.error("Error while retrieving organization discovery configuration.", e);
+            return false;
+        }
+    }
+
+    @Override
+    public int getPriority() {
+
+        int priority = super.getPriority();
+        if (priority == -1) {
+            priority = 15;
+        }
+        return priority;
+    }
+
+    @Override
+    public PostAuthnHandlerFlowStatus handle(HttpServletRequest request, HttpServletResponse response,
+                                             AuthenticationContext context) throws PostAuthenticationFailedException {
+
+        SequenceConfig sequenceConfig = context.getSequenceConfig();
+        for (Map.Entry<Integer, StepConfig> entry : sequenceConfig.getStepMap().entrySet()) {
+            StepConfig stepConfig = entry.getValue();
+            AuthenticatorConfig authenticatorConfig = stepConfig.getAuthenticatedAutenticator();
+            if (authenticatorConfig == null) {
+                continue;
+            }
+
+            ApplicationAuthenticator authenticator = authenticatorConfig.getApplicationAuthenticator();
+            if (authenticator instanceof FederatedApplicationAuthenticator) {
```

Comment:
Shall we test this with Email OTP/SMS OTP federated authenticator?

Reply:
As discussed offline this will impact the legacy email/sms otp authenticators if they are added to the login flow of the sub organizations. We can assume this won't be an issue since sub organization login flow management capabilities were not present when the legacy authenticators were used.